.ci/deps.sh

```diff
@@ -67,6 +67,12 @@ if [[ "x${PLUGIN}" == "xmodel/vowpalWabbit" ]] || [[ "x${PLUGIN}" == "x." ]] ||
   conda install -y -c conda-forge vowpalwabbit
 fi
 
+if [[ "x${PLUGIN}" == "xmodel/daal4py" ]] || [[ "x${PLUGIN}" == "x." ]] || [[ "x${PLUGIN}" == "xdocs" ]]; then
+  conda build conda-recipe -c intel -c conda-forge
```

Comment:
i tried to mimic the vowpal installation since its also being installed using conda. But daal4py needs recipe in order to be installed, which is available on their repo, and then build it and install using the file built in. Now I'm not quite certain how to address that.

Reply:
I'm not sure I understand. Where are you finding these installation instructions?

I haven't tried it, and it may be wrong, but https://intelpython.github.io/daal4py/#getting-daal4py points to anaconda where it says `conda install -c intel daal4py `

Reply:
Ah i see it's the building from source instructions, does replacing the three lines in this if block with `conda install -c intel daal4py` work?

Reply:
You've followed the format correctly :D

Reply:
Also, don't know if you saw this but we figured out how to run CI locally last weekend: https://intel.github.io/dffml/master/contributing/testing.html#running-ci-tests-locally

Reply:
Yeah I just pasted the format here so you don't have to look for it knowing that it would create errors lol. Guess I should've just commented it.

Reply:
> Also, don't know if you saw this but we figured out how to run CI locally last weekend: https://intel.github.io/dffml/master/contributing/testing.html#running-ci-tests-locally

Yeah I saw that but forgot about it. Thanks

Reply:
> i tried to mimic the vowpal installation since its also being installed using conda. But daal4py needs recipe in order to be installed, which is available on their repo, and then build it and install using the file built in. Now I'm not quite certain how to address that.

We have to figure out how to get the recipe, get its directory and build it. Also get the directory of the file it builds it in which was given in the logs. And use it to install daal4py PHEW! At least that's whats written on their repo and I installed it similarly.

Reply:
And this is the other way to do it https://github.com/IntelPython/daal4py#building-daal4py-without-conda-build

Reply:
Does `conda install -c intel daal4py` not work?

Reply:
conda seems to be producing errors when looking for some networking library, so we have to turn off exit on error (`set +e`)

```suggestion
  set +e
  conda build conda-recipe -c intel -c conda-forge
  echo "conda install -y -c intel daal4py, exit code: $?"
  set -e
```

Reply:
No it doesn't. Still requires recipe
`ValueError: No valid recipes found for input: ['conda-recipe']`
Is this working for you?

Reply:
Sorry, I put the wrong thing in the suggestion at first

Reply:
That worked thanks a lot!  Although I'm not quite sure why when they said the recipe is a requirement :thinking:

Reply:
@pdxjohnny 
Upon running the same docker command you just mentioned in the comment, ie. 

```
docker run --rm -ti -u $(id -u):$(id -g) \
    -e LOGGING=debug \
    -e USER=$USER \
    -v $HOME/.cache/pip:/home/$USER/.cache/pip -w /usr/src/dffml \
    -v $PWD:/usr/src/dffml -w /usr/src/dffml \
    --entrypoint .ci/docker-entrypoint.sh python:3.7 .
```

 I get the following log. There's no mention of daal4py being installed. I'm debugging this, just wanted to make sure that there weren't some other changes that you had made locally which dodged this. Posting the latter half of it which is installing the packages.

```
Successfully installed dffml
++ [[ x. == \x\f\e\a\t\u\r\e\/\g\i\t ]]
++ [[ x. == \x\o\p\e\r\a\t\i\o\n\s\/\d\e\p\l\o\y ]]
++ '[' . == source/mysql ']'
++ [[ x. == \x\m\o\d\e\l\/\v\o\w\p\a\l\W\a\b\b\i\t ]]
++ [[ x. == \x\. ]]
++ conda install -y -c conda-forge vowpalwabbit
++ '[' 5 -lt 1 ']'
++ local cmd=install
++ shift
++ case "$cmd" in
++ CONDA_INTERNAL_OLDPATH=/home/hashim/.cache/pip/minicondapy37/bin/libfabric:/home/hashim/.cache/pip/minicondapy37/bin:/home/hashim/.cache/pip/minicondapy37/condabin:/home/hashim/.cache/pip/minicondapy37/bin:/tmp/tmp.owxDQz06Lw/.local/bin:/tmp/tmp.JCXFyP09Gj/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
++ __add_sys_prefix_to_path
++ '[' -n '' ']'
+++ dirname /home/hashim/.cache/pip/minicondapy37/bin/conda
++ SYSP=/home/hashim/.cache/pip/minicondapy37/bin
+++ dirname /home/hashim/.cache/pip/minicondapy37/bin
++ SYSP=/home/hashim/.cache/pip/minicondapy37
++ '[' -n '' ']'
++ PATH=/home/hashim/.cache/pip/minicondapy37/bin:/home/hashim/.cache/pip/minicondapy37/bin/libfabric:/home/hashim/.cache/pip/minicondapy37/bin:/home/hashim/.cache/pip/minicondapy37/condabin:/home/hashim/.cache/pip/minicondapy37/bin:/tmp/tmp.owxDQz06Lw/.local/bin:/tmp/tmp.JCXFyP09Gj/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
++ export PATH
++ /home/hashim/.cache/pip/minicondapy37/bin/conda install -y -c conda-forge vowpalwabbit
Collecting package metadata (current_repodata.json): done
Solving environment: done

# All requested packages already installed.

++ local t1=0
++ PATH=/home/hashim/.cache/pip/minicondapy37/bin/libfabric:/home/hashim/.cache/pip/minicondapy37/bin:/home/hashim/.cache/pip/minicondapy37/condabin:/home/hashim/.cache/pip/minicondapy37/bin:/tmp/tmp.owxDQz06Lw/.local/bin:/tmp/tmp.JCXFyP09Gj/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
++ '[' 0 = 0 ']'
++ __conda_reactivate
++ local ask_conda
++ CONDA_INTERNAL_OLDPATH=/home/hashim/.cache/pip/minicondapy37/bin/libfabric:/home/hashim/.cache/pip/minicondapy37/bin:/home/hashim/.cache/pip/minicondapy37/condabin:/home/hashim/.cache/pip/minicondapy37/bin:/tmp/tmp.owxDQz06Lw/.local/bin:/tmp/tmp.JCXFyP09Gj/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
++ __add_sys_prefix_to_path
++ '[' -n '' ']'
+++ dirname /home/hashim/.cache/pip/minicondapy37/bin/conda
++ SYSP=/home/hashim/.cache/pip/minicondapy37/bin
+++ dirname /home/hashim/.cache/pip/minicondapy37/bin
++ SYSP=/home/hashim/.cache/pip/minicondapy37
++ '[' -n '' ']'
++ PATH=/home/hashim/.cache/pip/minicondapy37/bin:/home/hashim/.cache/pip/minicondapy37/bin/libfabric:/home/hashim/.cache/pip/minicondapy37/bin:/home/hashim/.cache/pip/minicondapy37/condabin:/home/hashim/.cache/pip/minicondapy37/bin:/tmp/tmp.owxDQz06Lw/.local/bin:/tmp/tmp.JCXFyP09Gj/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
++ export PATH
+++ PS1='(base) (tmp.JCXFyP09Gj) '
+++ /home/hashim/.cache/pip/minicondapy37/bin/conda shell.posix reactivate
++ ask_conda='. "/home/hashim/.cache/pip/minicondapy37/etc/conda/deactivate.d/mpivars.deactivate.sh"
PS1='\''(base) (tmp.JCXFyP09Gj) '\''
export PATH='\''/home/hashim/.cache/pip/minicondapy37/bin/libfabric:/home/hashim/.cache/pip/minicondapy37/bin:/home/hashim/.cache/pip/minicondapy37/condabin:/home/hashim/.cache/pip/minicondapy37/bin:/tmp/tmp.owxDQz06Lw/.local/bin:/tmp/tmp.JCXFyP09Gj/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'\''
export CONDA_SHLVL='\''1'\''
export CONDA_PROMPT_MODIFIER='\''(base) '\''
. "/home/hashim/.cache/pip/minicondapy37/etc/conda/activate.d/mpivars.activate.sh"'
++ PATH=/home/hashim/.cache/pip/minicondapy37/bin/libfabric:/home/hashim/.cache/pip/minicondapy37/bin:/home/hashim/.cache/pip/minicondapy37/condabin:/home/hashim/.cache/pip/minicondapy37/bin:/tmp/tmp.owxDQz06Lw/.local/bin:/tmp/tmp.JCXFyP09Gj/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
++ eval '. "/home/hashim/.cache/pip/minicondapy37/etc/conda/deactivate.d/mpivars.deactivate.sh"
PS1='\''(base) (tmp.JCXFyP09Gj) '\''
export PATH='\''/home/hashim/.cache/pip/minicondapy37/bin/libfabric:/home/hashim/.cache/pip/minicondapy37/bin:/home/hashim/.cache/pip/minicondapy37/condabin:/home/hashim/.cache/pip/minicondapy37/bin:/tmp/tmp.owxDQz06Lw/.local/bin:/tmp/tmp.JCXFyP09Gj/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'\''
export CONDA_SHLVL='\''1'\''
export CONDA_PROMPT_MODIFIER='\''(base) '\''
. "/home/hashim/.cache/pip/minicondapy37/etc/conda/activate.d/mpivars.activate.sh"'
+++ . /home/hashim/.cache/pip/minicondapy37/etc/conda/deactivate.d/mpivars.deactivate.sh
++++ '[' '' '!=' 1 ']'
+++++ echo /home/hashim/.cache/pip/minicondapy37/lib/mpi.jar
+++++ sed 's|/home/hashim/.cache/pip/minicondapy37/lib/mpi.jar:\?||'
++++ export CLASSPATH=
++++ CLASSPATH=
+++++ echo /home/hashim/.cache/pip/minicondapy37/lib/libfabric:
+++++ sed 's|/home/hashim/.cache/pip/minicondapy37/lib/libfabric:\?||'
++++ export LD_LIBRARY_PATH=
++++ LD_LIBRARY_PATH=
+++++ echo
+++++ sed 's|/home/hashim/.cache/pip/minicondapy37/lib:\?||'
++++ export LD_LIBRARY_PATH=
++++ LD_LIBRARY_PATH=
+++++ echo /home/hashim/.cache/pip/minicondapy37/share/man:
+++++ sed 's|/home/hashim/.cache/pip/minicondapy37/share/man:\?||'
++++ export MANPATH=
++++ MANPATH=
+++++ echo /home/hashim/.cache/pip/minicondapy37/lib/libfabric
+++++ sed 's|/home/hashim/.cache/pip/minicondapy37/lib/libfabric:\?||'
++++ export LIBRARY_PATH=
++++ LIBRARY_PATH=
+++++ echo /home/hashim/.cache/pip/minicondapy37/bin/libfabric:/home/hashim/.cache/pip/minicondapy37/bin:/home/hashim/.cache/pip/minicondapy37/condabin:/home/hashim/.cache/pip/minicondapy37/bin:/tmp/tmp.owxDQz06Lw/.local/bin:/tmp/tmp.JCXFyP09Gj/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+++++ sed 's|/home/hashim/.cache/pip/minicondapy37/bin/libfabric:||'
++++ export PATH=/home/hashim/.cache/pip/minicondapy37/bin:/home/hashim/.cache/pip/minicondapy37/condabin:/home/hashim/.cache/pip/minicondapy37/bin:/tmp/tmp.owxDQz06Lw/.local/bin:/tmp/tmp.JCXFyP09Gj/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
++++ PATH=/home/hashim/.cache/pip/minicondapy37/bin:/home/hashim/.cache/pip/minicondapy37/condabin:/home/hashim/.cache/pip/minicondapy37/bin:/tmp/tmp.owxDQz06Lw/.local/bin:/tmp/tmp.JCXFyP09Gj/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+++++ which fi_info
++++ FIP=

```

Reply:
The `set +e` before and `set -e` after are important here. It looks like conda was recently updated to look for a tool called `fi_info` which is from the recently released [libfabric](https://github.com/ofiwg/libfabric). Problem is it's looking for it in a way that can throw an error, and `set -e` mode means don't ignore errors, exit the script if anything errors, so we have to switch it to `set +e` mode, which allows / ignores errors. Errors being when commands exit with non zero return codes.

For example if we use grep to check for `hello`, but it only sees `hi`, it will exit with a non-zero return code (the return code of the previously run command can be accessed via the `$?` variable in bash)

```console
$ echo hi | grep hello
$ echo $?
1
```

```console
$ echo hello | grep hello
hello
$ echo $?
0
```

I'm just now realizing we touched on this a bit in today's meeting but for an entirely different reason: https://youtu.be/nn8_RJaMwto?t=487

More info: https://shapeshed.com/unix-exit-codes/

Reply:
@pdxjohnny  thanks for the detailed explanation. :) 
Although, I did set e+ and e- as you suggested previously but it didn't work thats why I asked you if you made any other changes. Some weird behavior here.... I guess I should push the code so you can get a better idea.